google-cloud-firestore/src/test/java/com/google/cloud/firestore/it/ITQueryTest.java

```diff
@@ -392,4 +389,480 @@ public void testOrderByEquality()
+        "Skip this test if running against production because multiple equality is "
+            + "not supported yet.",
+        isRunningAgainstFirestoreEmulator(firestore));
+
+    CollectionReference collection =
+        testCollectionWithDocs(
+            map(
+                "doc1",
+                map("key", "a", "sort", 0, "v", asList(0)),
+                "doc2",
+                map("key", "b", "sort", 1, "v", asList(0, 1, 3)),
+                "doc3",
+                map("key", "c", "sort", 1, "v", asList()),
+                "doc4",
+                map("key", "d", "sort", 2, "v", asList(1)),
+                "doc5",
+                map("key", "e", "sort", 3, "v", asList(2, 4)),
+                "doc6",
+                map("key", "f", "sort", 4, "v", Arrays.asList(Double.NaN)),
+                "doc7",
+                map("key", "g", "sort", 4, "v", Collections.singletonList(null))));
+
+    Query query1 =
+        collection
+            .whereNotEqualTo("key", "a")
+            .whereGreaterThanOrEqualTo("sort", 1)
+            .whereArrayContains("v", 0);
+    checkQuerySnapshotContainsDocuments(query1, "doc2");
+
+    Query query2 =
+        collection
+            .whereNotEqualTo("key", "a")
+            .whereGreaterThanOrEqualTo("sort", 1)
+            .whereArrayContainsAny("v", asList(0, 1));
+    checkQuerySnapshotContainsDocuments(query2, "doc2", "doc4");
+  }
+
+  private static Map<String, Object> nestedObject(int number) {
+    return map(
+        "name",
+        String.format("room %d", number),
+        "metadata",
+        map("createdAt", number),
+        "field",
+        String.format("field %d", number),
+        "field.dot",
+        number,
+        "field\\slash",
+        number);
+  }
+
+  // Use cursor in following test cases to add implicit order by fields in the sdk and compare the
+  // result with the query fields normalized in the server.
+  @Test
+  public void multipleInequalityWithNestedField() throws Exception {
+    // TODO(MIEQ): Enable this test against production when possible.
+    assumeTrue(
+        "Skip this test if running against production because multiple equality is "
+            + "not supported yet.",
+        isRunningAgainstFirestoreEmulator(firestore));
+
+    CollectionReference collection =
+        testCollectionWithDocs(
+            map(
+                "doc1", nestedObject(400),
+                "doc2", nestedObject(200),
+                "doc3", nestedObject(100),
+                "doc4", nestedObject(300)));
+
+    // ordered by: name asc, metadata.createdAt asc, __name__ asc
+    Query query1 =
+        collection
+            .whereLessThanOrEqualTo("metadata.createdAt", 500)
+            .whereGreaterThan("metadata.createdAt", 100)
+            .whereNotEqualTo("name", "room 200")
+            .orderBy("name");
+    DocumentSnapshot docSnap = collection.document("doc4").get().get();
+    Query query1_sdk_normalized = query1.startAt(docSnap);
```

Comment:
Shouldn't use a mix of snake case and camel case. 
query1_sdk_normalized --> query1SdkNormalized.

also, we can probably use a better name than "sdk normalized" . such as: query1WithCursor, query2WithCursor, etc.

The comment above (line 553) makes it clear that we're adding cursors to trigger the implicit orderBy logic.